File: clients/typescript/test/reactive.test.ts
import assert from "node:assert/strict";
import test from "node:test";

import { Subject } from "rxjs";

import {
  Client,
  ClientError,
  observeHealth,
  type ClientErrorCode,
  type HealthProbeState,
} from "../src/index.js";

const encoder = new TextEncoder();
const nextTurn = () => new Promise<void>((resolve) => setImmediate(resolve));

test("observeHealth emits immutable checking and available states", async () => {
  const triggers = new Subject<void>();
  const states: HealthProbeState[] = [];
  const client = new Client({
    baseUrl: "https://worker.example/",
    maxResponseBytes: 1024,
  });
  const subscription = observeHealth(triggers, client, async () =>
    encoder.encode('{"ok":true,"service":"gha-indie-worker-api-server"}'),
  ).subscribe((state) => states.push(state));

  triggers.next();
  await nextTurn();

  assert.deepEqual(
    states.map((state) => state.kind),
    ["checking", "available"],
  );
  assert.equal(states[1]?.endpoint, "https://worker.example/v1/health");
  assert.equal(Object.isFrozen(states[1]), true);
  subscription.unsubscribe();
});

test("a newer trigger aborts and suppresses the stale request", async () => {
  const triggers = new Subject<void>();
  const signals: AbortSignal[] = [];
  const resolutions: Array<(body: Uint8Array) => void> = [];
  const states: HealthProbeState[] = [];
  const client = new Client({ baseUrl: "https://worker.example", maxResponseBytes: 1024 });

  const subscription = observeHealth(
    triggers,
    client,
    (_endpoint, signal) =>
      new Promise<Uint8Array>((resolve) => {
        signals.push(signal);
        resolutions.push(resolve);
      }),
  ).subscribe((state) => states.push(state));

  triggers.next();
  triggers.next();
  assert.equal(signals[0]?.aborted, true);

  resolutions[0]?.(encoder.encode('{"ok":false,"service":"stale"}'));
  resolutions[1]?.(encoder.encode('{"ok":true,"service":"current"}'));
  await nextTurn();

  assert.equal(states.some((state) => state.kind === "available" && state.health.service === "stale"), false);
  assert.equal(states.at(-1)?.kind, "available");
  subscription.unsubscribe();
});

test("decode and transport failures become typed unavailable states", async () => {
  const triggers = new Subject<void>();
  const states: HealthProbeState[] = [];
  const client = new Client({ baseUrl: "https://worker.example", maxResponseBytes: 1024 });
  const subscription = observeHealth(triggers, client, async () => encoder.encode("{}"))
    .subscribe((state) => states.push(state));

  triggers.next();
  await nextTurn();

  const terminal = states.at(-1);
  assert.equal(terminal?.kind, "unavailable");
  if (terminal?.kind === "unavailable") {
    assert.equal(terminal.error instanceof ClientError, true);
    assert.equal(terminal.error.code, "invalid_shape");
  }
  subscription.unsubscribe();
});

test("malformed, oversized, and transport failures retain typed error codes", async () => {
  const client = new Client({
    baseUrl: "https://worker.example",
    maxResponseBytes: 8,
  });
  const cases: ReadonlyArray<
    readonly [string, () => Promise<Uint8Array>, ClientErrorCode]
  > = [
    ["malformed JSON", async () => encoder.encode("{"), "invalid_json"],
    ["oversized body", async () => encoder.encode("123456789"), "too_large"],
    [
      "transport rejection",
      async () => Promise.reject(new Error("connection reset")),
      "http",
    ],
  ];

  for (const [name, request, expectedCode] of cases) {
    const triggers = new Subject<void>();
    const states: HealthProbeState[] = [];
    const subscription = observeHealth(triggers, client, request).subscribe(
      (state) => states.push(state),
    );

    triggers.next();
    await nextTurn();

    const terminal = states.at(-1);
    assert.equal(terminal?.kind, "unavailable", name);
    if (terminal?.kind === "unavailable") {
      assert.equal(terminal.error.code, expectedCode, name);
    }
    subscription.unsubscribe();
    triggers.complete();
  }
});

test("concurrent and late subscribers share one request and replay its result", async () => {
  const triggers = new Subject<void>();
  const client = new Client({
    baseUrl: "https://worker.example",
    maxResponseBytes: 1024,
  });
  let requestCount = 0;
  let resolveRequest: ((body: Uint8Array) => void) | undefined;
  const statesBySubscriber: HealthProbeState[][] = [[], [], []];
  const stream = observeHealth(triggers, client, async () => {
    requestCount += 1;
    return new Promise<Uint8Array>((resolve) => {
      resolveRequest = resolve;
    });
  });

  const first = stream.subscribe((state) => statesBySubscriber[0]?.push(state));
  const second = stream.subscribe((state) => statesBySubscriber[1]?.push(state));
  triggers.next();
  assert.equal(requestCount, 1);

  resolveRequest?.(
    encoder.encode('{"ok":true,"service":"shared-result"}'),
  );
  await nextTurn();

  const late = stream.subscribe((state) => statesBySubscriber[2]?.push(state));
  assert.equal(requestCount, 1);
  assert.deepEqual(
    statesBySubscriber.map((states) => states.at(-1)?.kind),
    ["available", "available", "available"],
  );

  first.unsubscribe();
  second.unsubscribe();
  late.unsubscribe();
  triggers.complete();
});

test("unsubscribing the last observer aborts the active request", () => {
  const triggers = new Subject<void>();
  const client = new Client({
    baseUrl: "https://worker.example",
    maxResponseBytes: 1024,
  });
  let requestSignal: AbortSignal | undefined;
  const subscription = observeHealth(
    triggers,
    client,
    (_endpoint, signal) => {
      requestSignal = signal;
      return new Promise<Uint8Array>(() => undefined);
    },
  ).subscribe();

  triggers.next();
  assert.equal(requestSignal?.aborted, false);
  subscription.unsubscribe();
  assert.equal(requestSignal?.aborted, true);
  triggers.complete();
});

test("the client snapshots mutable constructor input", () => {
  const config = {
    baseUrl: "https://worker.example/original",
    maxResponseBytes: 1024,
  };
  const client = new Client(config);

  config.baseUrl = "https://attacker.example/replaced";

  assert.equal(
    client.healthUrl(),
    "https://worker.example/original/v1/health",
  );
});

File: clients/dart/test/reactive_test.dart
import 'dart:async';
import 'dart:convert';
import 'dart:typed_data';

import 'package:gha_indie_worker_client/gha_indie_worker_client.dart';
import 'package:test/test.dart';

void main() {
  test('emits checking and available as immutable state variants', () async {
    final triggers = StreamController<void>();
    final client =
        Client(const ClientConfig(baseUrl: 'https://worker.example/'));
    final states = <HealthProbeState>[];
    final subscription = observeHealth(
      triggers: triggers.stream,
      client: client,
      request: (_) => Stream.value(
        Uint8List.fromList(utf8.encode('{"ok":true,"service":"api"}')),
      ),
    ).listen(states.add);

    triggers.add(null);
    await Future<void>.delayed(Duration.zero);

    expect(states, [isA<HealthChecking>(), isA<HealthAvailable>()]);
    expect((states.last as HealthAvailable).health.service, 'api');
    await subscription.cancel();
    await triggers.close();
  });

  test('switchMap cancels a stale request stream', () async {
    final triggers = StreamController<void>();
    final requests = <StreamController<Uint8List>>[];
    var cancellations = 0;
    final client =
        Client(const ClientConfig(baseUrl: 'https://worker.example'));
    final subscription = observeHealth(
      triggers: triggers.stream,
      client: client,
      request: (_) {
        late StreamController<Uint8List> controller;
        controller = StreamController<Uint8List>(
          onCancel: () {
            cancellations += 1;
          },
        );
        requests.add(controller);
        return controller.stream;
      },
    ).listen((_) {});

    triggers.add(null);
    await Future<void>.delayed(Duration.zero);
    triggers.add(null);
    await Future<void>.delayed(Duration.zero);

    expect(cancellations, 1);
    await requests.last.close();
    await subscription.cancel();
    await triggers.close();
  });

  test('decode and transport failures become typed unavailable states',
      () async {
    final cases =
        <({String name, HealthRequest request, ClientErrorCode code})>[
      (
        name: 'malformed JSON',
        request: (_) => Stream.value(Uint8List.fromList(utf8.encode('{'))),
        code: ClientErrorCode.invalidJson,
      ),
      (
        name: 'oversized body',
        request: (_) =>
            Stream.value(Uint8List.fromList(utf8.encode('123456789'))),
        code: ClientErrorCode.tooLarge,
      ),
      (
        name: 'transport error',
        request: (_) => Stream.error(StateError('connection reset')),
        code: ClientErrorCode.http,
      ),
    ];
    final client = Client(
      const ClientConfig(
        baseUrl: 'https://worker.example',
        maxResponseBytes: 8,
      ),
    );

    for (final testCase in cases) {
      final states = await observeHealth(
        triggers: Stream.value(null),
        client: client,
        request: testCase.request,
      ).toList();

      expect(states.last, isA<HealthUnavailable>(), reason: testCase.name);
      expect(
        (states.last as HealthUnavailable).error.code,
        testCase.code,
        reason: testCase.name,
      );
    }
  });

  test('concurrent and late subscribers share and replay one request',
      () async {
    final triggers = StreamController<void>.broadcast();
    final response = StreamController<Uint8List>();
    final client =
        Client(const ClientConfig(baseUrl: 'https://worker.example'));
    var requestCount = 0;
    final stream = observeHealth(
      triggers: triggers.stream,
      client: client,
      request: (_) {
        requestCount += 1;
        return response.stream;
      },
    );
    final firstStates = <HealthProbeState>[];
    final secondStates = <HealthProbeState>[];
    final lateStates = <HealthProbeState>[];
    final first = stream.listen(firstStates.add);
    final second = stream.listen(secondStates.add);

    triggers.add(null);
    await Future<void>.delayed(Duration.zero);
    expect(requestCount, 1);
    response.add(
      Uint8List.fromList(
        utf8.encode('{"ok":true,"service":"shared-result"}'),
      ),
    );
    await Future<void>.delayed(Duration.zero);

    final late = stream.listen(lateStates.add);
    await Future<void>.delayed(Duration.zero);
    expect(requestCount, 1);
    expect(firstStates.last, isA<HealthAvailable>());
    expect(secondStates.last, isA<HealthAvailable>());
    expect(lateStates.last, isA<HealthAvailable>());

    await first.cancel();
    await second.cancel();
    await late.cancel();
    await response.close();
    await triggers.close();
  });

  test('a later trigger recovers after an unavailable state', () async {
    final triggers = StreamController<void>.broadcast(sync: true);
    final states = <HealthProbeState>[];
    final client =
        Client(const ClientConfig(baseUrl: 'https://worker.example'));
    var requestCount = 0;
    final subscription = observeHealth(
      triggers: triggers.stream,
      client: client,
      request: (_) {
        requestCount += 1;
        if (requestCount == 1) {
          return Stream.error(StateError('temporary outage'));
        }
        return Stream.value(
          Uint8List.fromList(
            utf8.encode('{"ok":true,"service":"recovered-api"}'),
          ),
        );
      },
    ).listen(states.add);

    triggers.add(null);
    await Future<void>.delayed(Duration.zero);
    expect(states.last, isA<HealthUnavailable>());

    triggers.add(null);
    await Future<void>.delayed(Duration.zero);
    expect(requestCount, 2);
    expect(states.last, isA<HealthAvailable>());

    await subscription.cancel();
    await triggers.close();
  });
}
